File: go/internal/e2e/multi_turn_e2e_test.go
package e2e

import (
	"os"
	"path/filepath"
	"strings"
	"sync"
	"testing"

	copilot "github.com/github/copilot-sdk/go"
	"github.com/github/copilot-sdk/go/internal/e2e/testharness"
)

func TestMultiTurnE2E(t *testing.T) {
	ctx := testharness.NewTestContext(t)
	client := ctx.NewClient()
	t.Cleanup(func() { client.ForceStop() })

	t.Run("should use tool results from previous turns", func(t *testing.T) {
		ctx.ConfigureForTest(t)

		if err := os.WriteFile(filepath.Join(ctx.WorkDir, "secret.txt"), []byte("The magic number is 42."), 0644); err != nil {
			t.Fatalf("Failed to write secret.txt: %v", err)
		}

		session, err := client.CreateSession(t.Context(), &copilot.SessionConfig{
			OnPermissionRequest: copilot.PermissionHandler.ApproveAll,
		})
		if err != nil {
			t.Fatalf("Failed to create session: %v", err)
		}
		t.Cleanup(func() { _ = session.Disconnect() })

		var mu sync.Mutex
		var events []copilot.SessionEvent
		session.On(func(event copilot.SessionEvent) {
			mu.Lock()
			events = append(events, event)
			mu.Unlock()
		})

		msg1, err := session.SendAndWait(t.Context(), copilot.MessageOptions{
			Prompt: "Read the file 'secret.txt' and tell me what the magic number is.",
		})
		if err != nil {
			t.Fatalf("First SendAndWait failed: %v", err)
		}
		if content := assistantContent(t, msg1); !strings.Contains(content, "42") {
			t.Fatalf("Expected first response to contain 42, got %q", content)
		}
		assertToolTurnOrdering(t, snapshotAndClearMultiTurnEvents(&mu, &events), "file read turn")

		msg2, err := session.SendAndWait(t.Context(), copilot.MessageOptions{
			Prompt: "What is that magic number multiplied by 2?",
		})
		if err != nil {
			t.Fatalf("Second SendAndWait failed: %v", err)
		}
		if content := assistantContent(t, msg2); !strings.Contains(content, "84") {
			t.Fatalf("Expected second response to contain 84, got %q", content)
		}
	})

	t.Run("should handle file creation then reading across turns", func(t *testing.T) {
		ctx.ConfigureForTest(t)

		session, err := client.CreateSession(t.Context(), &copilot.SessionConfig{
			OnPermissionRequest: copilot.PermissionHandler.ApproveAll,
		})
		if err != nil {
			t.Fatalf("Failed to create session: %v", err)
		}
		t.Cleanup(func() { _ = session.Disconnect() })

		var mu sync.Mutex
		var events []copilot.SessionEvent
		session.On(func(event copilot.SessionEvent) {
			mu.Lock()
			events = append(events, event)
			mu.Unlock()
		})

		if _, err := session.SendAndWait(t.Context(), copilot.MessageOptions{
			Prompt: "Create a file called 'greeting.txt' with the content 'Hello from multi-turn test'.",
		}); err != nil {
			t.Fatalf("First SendAndWait failed: %v", err)
		}
		// File should have been created with the expected content
		greetingContent, err := os.ReadFile(filepath.Join(ctx.WorkDir, "greeting.txt"))
		if err != nil {
			t.Fatalf("Failed to read greeting.txt: %v", err)
		}
		if !strings.Contains(string(greetingContent), "Hello from multi-turn test") {
			t.Errorf("Expected greeting.txt to contain 'Hello from multi-turn test', got %q", string(greetingContent))
		}
		assertToolTurnOrdering(t, snapshotAndClearMultiTurnEvents(&mu, &events), "file creation turn")

		msg, err := session.SendAndWait(t.Context(), copilot.MessageOptions{
			Prompt: "Read the file 'greeting.txt' and tell me its exact contents.",
		})
		if err != nil {
			t.Fatalf("Second SendAndWait failed: %v", err)
		}
		if content := assistantContent(t, msg); !strings.Contains(content, "Hello from multi-turn test") {
			t.Fatalf("Expected response to contain created file contents, got %q", content)
		}
		assertToolTurnOrdering(t, snapshotAndClearMultiTurnEvents(&mu, &events), "file read turn")
	})
}

func snapshotAndClearMultiTurnEvents(mu *sync.Mutex, events *[]copilot.SessionEvent) []copilot.SessionEvent {
	mu.Lock()
	defer mu.Unlock()
	snapshot := make([]copilot.SessionEvent, len(*events))
	copy(snapshot, *events)
	*events = (*events)[:0]
	return snapshot
}

// assertToolTurnOrdering verifies that for a turn with tool use the events arrive in the
// expected order: user.message → tool.execution_start(s) → tool.execution_complete(s)
// → assistant.message → session.idle.
func assertToolTurnOrdering(t *testing.T, events []copilot.SessionEvent, turnDescription string) {
	t.Helper()

	observedTypes := make([]copilot.SessionEventType, 0, len(events))
	for _, e := range events {
		observedTypes = append(observedTypes, e.Type())
	}

	userMessageIdx := indexOfEventType(events, copilot.SessionEventTypeUserMessage, 0)
	if userMessageIdx < 0 {
		// A turn without a tool call (e.g., pure text answer) may not need ordering.
		// Only assert if tool events are present.
		if !containsEventType(events, copilot.SessionEventTypeToolExecutionStart) {
			return
		}
		t.Errorf("Expected user.message in %s but none found; types=%v", turnDescription, observedTypes)
		return
	}

	firstToolStartIdx := indexOfEventType(events, copilot.SessionEventTypeToolExecutionStart, 0)
	if firstToolStartIdx < 0 {
		// No tool use in this turn — nothing to assert.
		return
	}
	lastToolCompleteIdx := lastIndexOfEventType(events, copilot.SessionEventTypeToolExecutionComplete)
	assistantAfterToolsIdx := indexOfEventType(events, copilot.SessionEventTypeAssistantMessage, lastToolCompleteIdx+1)
	sessionIdleIdx := indexOfEventType(events, copilot.SessionEventTypeSessionIdle, 0)

	if userMessageIdx >= firstToolStartIdx {
		t.Errorf("[%s] Expected user.message before first tool start; types=%v", turnDescription, observedTypes)
	}

	// Match each tool.execution_complete to a preceding tool.execution_start with the same ToolCallID.
	starts := make(map[string]int)
	for i, e := range events {
		if e.Type() == copilot.SessionEventTypeToolExecutionStart {
			if d, ok := e.Data.(*copilot.ToolExecutionStartData); ok {
				starts[d.ToolCallID] = i
			}
		}
	}
	for _, e := range events {
		if e.Type() == copilot.SessionEventTypeToolExecutionComplete {
			if d, ok := e.Data.(*copilot.ToolExecutionCompleteData); ok {
				if _, found := starts[d.ToolCallID]; !found {
					t.Errorf("[%s] tool.execution_complete for %q has no matching tool.execution_start; types=%v",
						turnDescription, d.ToolCallID, observedTypes)
				}
			}
		}
	}

	if assistantAfterToolsIdx < 0 {
		t.Errorf("[%s] Expected assistant.message after final tool completion; types=%v", turnDescription, observedTypes)
	}
	if sessionIdleIdx < 0 {
		t.Errorf("[%s] Expected session.idle; types=%v", turnDescription, observedTypes)
	}
	if assistantAfterToolsIdx >= 0 && lastToolCompleteIdx >= assistantAfterToolsIdx {
		t.Errorf("[%s] Expected final tool completion before final assistant.message; types=%v", turnDescription, observedTypes)
	}
	if assistantAfterToolsIdx >= 0 && sessionIdleIdx >= 0 && assistantAfterToolsIdx >= sessionIdleIdx {
		t.Errorf("[%s] Expected assistant.message before session.idle; types=%v", turnDescription, observedTypes)
	}
}

func indexOfEventType(events []copilot.SessionEvent, typ copilot.SessionEventType, startIdx int) int {
	for i := startIdx; i < len(events); i++ {
		if events[i].Type() == typ {
			return i
		}
	}
	return -1
}

func lastIndexOfEventType(events []copilot.SessionEvent, typ copilot.SessionEventType) int {
	for i := len(events) - 1; i >= 0; i-- {
		if events[i].Type() == typ {
			return i
		}
	}
	return -1
}

func containsEventType(events []copilot.SessionEvent, typ copilot.SessionEventType) bool {
	return indexOfEventType(events, typ, 0) >= 0
}

File: nodejs/test/e2e/multi_turn.e2e.test.ts
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *--------------------------------------------------------------------------------------------*/

import { writeFile } from "fs/promises";
import { join } from "path";
import { describe, expect, it } from "vitest";
import { SessionEvent, approveAll } from "../../src/index.js";
import { createSdkTestContext } from "./harness/sdkTestContext";

describe("Multi-turn Tool Usage", async () => {
    const { copilotClient: client, workDir } = await createSdkTestContext();

    function snapshotAndClearEvents(events: SessionEvent[]): SessionEvent[] {
        const snapshot = [...events];
        events.length = 0;
        return snapshot;
    }

    function assertToolTurnOrdering(turnEvents: SessionEvent[], turnDescription: string): void {
        const types = turnEvents.map((e) => e.type);
        const observedTypes = types.join(", ");

        const userMsgIdx = types.indexOf("user.message");
        expect(
            userMsgIdx,
            `Expected user.message in ${turnDescription}. Observed: ${observedTypes}`
        ).toBeGreaterThanOrEqual(0);

        const toolStarts = turnEvents
            .map((e, i) => ({ e, i }))
            .filter(({ e }) => e.type === "tool.execution_start");
        const toolCompletes = turnEvents
            .map((e, i) => ({ e, i }))
            .filter(({ e }) => e.type === "tool.execution_complete");

        expect(
            toolStarts.length,
            `Expected tool starts in ${turnDescription}. Observed: ${observedTypes}`
        ).toBeGreaterThan(0);
        expect(
            toolCompletes.length,
            `Expected tool completes in ${turnDescription}. Observed: ${observedTypes}`
        ).toBeGreaterThan(0);

        const firstToolStartIdx = Math.min(...toolStarts.map(({ i }) => i));
        expect(
            userMsgIdx,
            `Expected user.message before first tool start in ${turnDescription}. Observed: ${observedTypes}`
        ).toBeLessThan(firstToolStartIdx);

        for (const { e: complete, i: completeIdx } of toolCompletes) {
            const matchingStart = toolStarts.find(
                ({ e: start, i: startIdx }) =>
                    start.data.toolCallId === complete.data.toolCallId && startIdx < completeIdx
            );
            expect(
                matchingStart,
                `Expected matching tool start for tool complete with id ${complete.data.toolCallId}`
            ).toBeDefined();
        }

        const lastToolCompleteIdx = Math.max(...toolCompletes.map(({ i }) => i));
        let assistantAfterToolsIdx = -1;
        for (let i = lastToolCompleteIdx + 1; i < turnEvents.length; i++) {
            if (turnEvents[i]!.type === "assistant.message") {
                assistantAfterToolsIdx = i;
                break;
            }
        }

        let sessionIdleIdx = -1;
        const searchFrom = assistantAfterToolsIdx >= 0 ? assistantAfterToolsIdx + 1 : 0;
        for (let i = searchFrom; i < turnEvents.length; i++) {
            if (turnEvents[i]!.type === "session.idle") {
                sessionIdleIdx = i;
                break;
            }
        }

        expect(
            assistantAfterToolsIdx,
            `Expected assistant.message after tool completion in ${turnDescription}. Observed: ${observedTypes}`
        ).toBeGreaterThanOrEqual(0);
        expect(
            sessionIdleIdx,
            `Expected session.idle after assistant.message in ${turnDescription}. Observed: ${observedTypes}`
        ).toBeGreaterThanOrEqual(0);
        expect(
            lastToolCompleteIdx,
            `Expected final tool completion before final assistant message in ${turnDescription}. Observed: ${observedTypes}`
        ).toBeLessThan(assistantAfterToolsIdx);
        expect(
            assistantAfterToolsIdx,
            `Expected final assistant message before idle in ${turnDescription}. Observed: ${observedTypes}`
        ).toBeLessThan(sessionIdleIdx);
    }

    it("should use tool results from previous turns", async () => {
        // Write a file, then ask the model to read it and reason about its content
        await writeFile(join(workDir, "secret.txt"), "The magic number is 42.");
        const session = await client.createSession({ onPermissionRequest: approveAll });
        const events: SessionEvent[] = [];
        session.on((event) => {
            events.push(event);
        });

        const msg1 = await session.sendAndWait({
            prompt: "Read the file 'secret.txt' and tell me what the magic number is.",
        });
        expect(msg1?.data.content).toContain("42");
        assertToolTurnOrdering(snapshotAndClearEvents(events), "file read turn");

        // Follow-up that requires context from the previous turn
        const msg2 = await session.sendAndWait({
            prompt: "What is that magic number multiplied by 2?",
        });
        expect(msg2?.data.content).toContain("84");
    });

    it("should handle file creation then reading across turns", async () => {
        const session = await client.createSession({ onPermissionRequest: approveAll });
        const events: SessionEvent[] = [];
        session.on((event) => {
            events.push(event);
        });

        // First turn: create a file
        await session.sendAndWait({
            prompt: "Create a file called 'greeting.txt' with the content 'Hello from multi-turn test'.",
        });

        // Verify file was created with correct content before checking ordering
        const { readFile } = await import("fs/promises");
        const createdContent = await readFile(join(workDir, "greeting.txt"), "utf-8");
        expect(createdContent).toBe("Hello from multi-turn test");
        assertToolTurnOrdering(snapshotAndClearEvents(events), "file creation turn");

        // Second turn: read the file
        const msg = await session.sendAndWait({
            prompt: "Read the file 'greeting.txt' and tell me its exact contents.",
        });
        expect(msg?.data.content).toContain("Hello from multi-turn test");
        assertToolTurnOrdering(snapshotAndClearEvents(events), "file read turn");
    });
});
